### PR TITLE
[SPARK-39419][SQL] Fix ArraySort to throw an exception when the comparator returns null

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Fixes `ArraySort` to throw an exception when the comparator returns `null`.

Also updates the doc to follow the corrected behavior.

### Why are the changes needed?

When the comparator of `ArraySort` returns `null`, currently it handles it as `0` (equal).

According to the doc, 

```
It returns -1, 0, or 1 as the first element is less than, equal to, or greater than
the second element. If the comparator function returns other
values (including null), the function will fail and raise an error.
```

It's fine to return non -1, 0, 1 integers to follow the Java convention (still need to update the doc, though), but it should throw an exception for `null` result.

### Does this PR introduce _any_ user-facing change?

Yes, if a user uses a comparator that returns `null`, it will throw an error after this PR.

The legacy flag `spark.sql.legacy.allowNullComparisonResultInArraySort` can be used to restore the legacy behavior that handles `null` as `0` (equal).

### How was this patch tested?

Added some tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -3828,6 +3828,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val LEGACY_ARRAY_SORT_HANDLES_COMPARISON_RESULT_NULL_VALUE_AS_ZERO =
+    buildConf("spark.sql.legacy.arraySortHandlesComparisonResultNullAsZero")
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3832]: qq should it be covered by ansi config @gengliangwang ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 2]
```diff
@@ -2012,4 +2012,12 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase {
     new SparkException(
       errorClass = "MULTI_VALUE_SUBQUERY_ERROR", messageParameters = Array(plan), cause = null)
   }
+
+  def comparisonResultIsNullError(): Throwable = {
+    new NullPointerException(
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:2017]: I remember the exception being thrown should inherit `SparkThrowable` when it's thrown to end-users ... cc @MaxGekk am I remembering it correct?

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise. two questions

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:2017]: yea, with the new error class framework, the java exception type doesn't matter any more. We can always use `SparkException`, but we need to add an error class for it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3832]: I don't think so, legacy config is for legacy behaviors.

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3832]: +1 with @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala [hunk 3]
```diff
@@ -375,9 +375,17 @@ case class ArrayTransform(
 // scalastyle:on line.size.limit
 case class ArraySort(
     argument: Expression,
-    function: Expression)
+    function: Expression,
+    handleComparisonResultNullAsZero: Boolean)
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala:379]: How about `failOnNullComparisonResult`?

**Review by gengliangwang — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -3828,6 +3828,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val LEGACY_ARRAY_SORT_FAILS_ON_NULL_COMPARISON_RESULT =
+    buildConf("spark.sql.legacy.arraySortFailsOnNullComparisonResult")
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3832]: I feel the "fail" behavior is the legacy one from the new conf name. The previous name was good. Or we can follow the other conf style `spark.sql.legacy.allowNullComparisonResultInArraySort`.
Sorry about the previous trivial comment: https://github.com/apache/spark/pull/36812#discussion_r892996318

File: core/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -551,5 +551,10 @@
       "Writing job aborted"
     ],
     "sqlState" : "40000"
+  },
+  "NULL_COMPARISON_RESULT" : {
+    "message" : [
+      "The comparison result is null. If you want to handle null as 0 (equal), you can set \"<config>\" to \"<value>\"."
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:557]: actually, why not hardcode the config name and value here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 4 above)
Comment (ueshin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3832]: No worries, thanks for the suggestions!
`spark.sql.legacy.allowNullComparisonResultInArraySort` sounds better to me.

File: core/src/main/resources/error/error-classes.json (hunk 5 above)
Comment (ueshin) [reply core/src/main/resources/error/error-classes.json:557]: I just thought it could be reusable. Let me just change to hardcode it for now.

**Review by gengliangwang — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
@ueshin please fix the conflicts, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -3818,6 +3818,16 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val LEGACY_ALLOW_NULL_COMPARISON_RESULT_IN_ARRAY_SORT =
+    buildConf("spark.sql.legacy.allowNullComparisonResultInArraySort")
+      .internal()
+      .doc("When set to false, `array_sort` function throws an error " +
+        "if the comparator function returns null. " +
+        "If set to true, it restores the legacy behavior that handles null as zero (equal).")
+      .version("3.2.2")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3827]: cc @MaxGekk FYI

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3827]: This is not a 3.3 regression so technically not a release blocker.

**Comment by cloud-fan** [comment]
thanks, merging to master! @ueshin can you open backport PRs for 3.2 and 3.3?

**Pull request closed without merging**